doc/spec/json.md

```diff
@@ -0,0 +1,92 @@
+# Docker Distribution JSON Canonicalization
+
+## Introduction
+
+To provide consistent content hashing of JSON object throughout Docker
+Distribution APIs, a canonical JSON format has been defined. The
+canonicalization is balance between conistency and readability. The format
+also provides consistent output for effective caching.
+
+## Rules
+
+Compliant JSON should conform to the following rules:
+
+1. All generated JSON should comply with [RFC
+   7159](http://www.ietf.org/rfc/rfc7159.txt).
+2. Resulting "JSON text" shall always be encoded in UTF-8.
+3. Unless a canonical key order is defined for a particular schema, object
+   keys shall always appear in lexically sorted order.
+4. The colon (":") shall always be followed by a single space.
+5. An indent of three spaces ("   ") shall be used.
```

Comment:
I realize that readability seems a big deal, but for APIs, this is an efficiency and streaming concern. I _love_ when JSON structures, in their entirety, are delimited by newlines. Then you can stream them, and marshal a line at a time.

As soon as you have the whole structure dependent on an EOF, then you have to buffer the whole stream, and marshal it. This may not be a problem in your initial use case, but having the canonical format as described will block future possible efficiencies, like streaming of json.